apps/extension/src/routes/popup/approval/tx-msg-handler.ts

```diff
@@ -4,19 +4,38 @@ import { PopupRequest } from '@penumbra-zone/types/src/internal-msg/popup';
 import { useStore } from '../../../state';
 import { AuthorizeRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/custody/v1alpha1/custody_pb';
 import { TransactionView } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import {
+  asPublicTransactionView,
+  asReceiverTransactionView,
+  classifyTransaction,
+} from '@penumbra-zone/types';
+import { isControlledAddress } from './is-controlled-address';
 
 export const isTxApprovalReq = (req: PopupRequest): req is TxApproval => {
   return req.type === 'TX-APPROVAL';
 };
 
 export const handleTxApproval: InternalMessageHandler<TxApproval> = (jsonReq, responder) => {
-  useStore.setState(state => {
-    state.txApproval.authorizeRequest = AuthorizeRequest.fromJson(
-      jsonReq.authorizeRequest,
-    ).toJsonString();
-    state.txApproval.transactionViewFromPlan = TransactionView.fromJson(
-      jsonReq.transactionViewFromPlan,
+  const authorizeRequestAsString = AuthorizeRequest.fromJson(
+    jsonReq.authorizeRequest,
+  ).toJsonString();
+
+  const transactionView = TransactionView.fromJson(jsonReq.transactionViewFromPlan);
+
+  void (async () => {
+    const asSender = transactionView.toJsonString();
+    const asPublic = asPublicTransactionView(transactionView).toJsonString();
+    const asReceiver = (
+      await asReceiverTransactionView(transactionView, { isControlledAddress })
     ).toJsonString();
-    state.txApproval.responder = responder;
-  });
+
+    useStore.setState(state => {
+      state.txApproval.authorizeRequest = authorizeRequestAsString;
+      state.txApproval.asSender = asSender;
+      state.txApproval.asReceiver = asReceiver;
+      state.txApproval.asPublic = asPublic;
+      state.txApproval.transactionClassification = classifyTransaction(transactionView);
+      state.txApproval.responder = responder;
```

Comment:
Hear me out. What if we kill nearly everything in `TxApproval` except for setting the `responder` in zustand:

https://github.com/penumbra-zone/web/blob/7d718d785dbc6fd0c63a5111d7f1b3c94706f648/apps/extension/src/routes/popup/approval/tx-msg-handler.ts#L20

Instead, pass all tx data as query params in the request URL:

https://github.com/penumbra-zone/web/blob/7d718d785dbc6fd0c63a5111d7f1b3c94706f648/packages/router/src/ctx/approver.ts#L11

We then kill all react hooks, handle data fetching in the react-router loader for that route, and the components have the data they need to render when they load. No more hooks+async required.

Reply:
In short, yes. Is it really _that_ big? If so, I suppose that is a hard blocker. If not, I think it's worth it. Url-as-state-management is basically the philosophy that gave rise to remixjs. It allows you to exit async react-hook land and prevents re-request waterfalls of re-render->request->re-render->request (I don't think this is the case for us here though).

Reply:
Just asked chatgpt for an example json object of 2000 characters. It's smaller than I expected lol. Looks like it's too close to call 😢

Reply:
Hmm: https://chromium.googlesource.com/chromium/src/+/master/docs/security/url_display_guidelines/url_display_guidelines.md#:~:text=URL%20Length,-In%20general%2C%20the&text=Chrome%20limits%20URLs%20to%20a,is%20used%20on%20VR%20platforms.

Not sure if this applies to chrome extension runtime though 🤔

Reply:
OK! Interesting. I mean, we could test it. Doesn't it just _feel_ kinda wrong, though, to use the URL? Not that that's a great reason not to do it. It just feels like there must be a better way...